### PR TITLE
Use SLEEF functions for NEON vectors on macOS ARM64

### PR DESCRIPTION
**Pull request by danieldk**
We noticed that on M1 Macs Tranformer network profiles are dominated by scalar `exp` and `erff` functions (for softmax and GELU).

The NEON `Vectorized<float>` implementation does not use SLEEF functions in order to compile on mobile platforms. However, SLEEF is already compiled on macOS ARM64 and is safe to use there. This change adds another implementation of `Vectorized<float>` that uses SLEEF functions. This implementation is only used on macOS ARM64.

This change speeds up e.g. prediction of spaCy transformer models by 20% on M1 Macs.

### CODE REVIEW

**Comment by danieldk** [comment]
What can I do to move this PR forward? It seems that the CI failure is a just a timeout:

```
Download action repository 'seemethere/add-github-ssh-key@v1' (SHA:40117388ae0680b2b1809aa182a68b87cc8c2e6a)
24
Warning: Failed to download action 'https://api.github.com/repos/seemethere/add-github-ssh-key/tarball/40117388ae0680b2b1809aa182a68b87cc8c2e6a'. Error: The request was canceled due to the configured HttpClient.Timeout of 100 seconds elapsing.
```

**Comment by kimishpatel** [comment]
> This change speeds up e.g. prediction of spaCy transformer models by 20% on M1 Macs.

@danieldk sorry for the late response. I will review this. @malfet for CI issue.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
At a high level: Can you not change vec256_float_neon.h to ifdef exp and errf?

**Comment by danieldk** [comment]
> At a high level: Can you not change vec256_float_neon.h to ifdef exp and errf?

I thought it'd be best to provide SLEEF implementations for all methods like x86_64. But of course, only conditionally using SLEEF for `exp` and `erf` is lower-impact. I'll make a separate PR to still have this PR in case someone wants to use SLEEF for the other methods in the future.

**Comment by danieldk** [comment]
> At a high level: Can you not change vec256_float_neon.h to ifdef exp and errf?

Created a new PR #71302 with this suggested change.

**Comment by kimishpatel** [comment]
> > At a high level: Can you not change vec256_float_neon.h to ifdef exp and errf?
> 
> I thought it'd be best to provide SLEEF implementations for most methods like x86_64. But only conditionally using SLEEF for `exp` and `erf` is lower-impact. I'll make a separate PR to still have this PR in case someone wants to use SLEEF for the other methods in the future.

I think it is fine to add support for all and not just exp and erf. I was just suggesting to do that as part of the same class for code reusability since we are using sleep only for transcendental ops.

**Comment by danieldk** [comment]
Ah, I see, on it!

**Comment by danieldk** [comment]
Done. I am not sure if you prefer using `if defined` for every method or for the whole bunch of methods. I now did the latter, I grouped the methods that can be implemented using SLEEF together and used a single `if defined ... else` block.

**Comment by kimishpatel** [comment]
> Done. I am not sure if you prefer using `if defined` for every method or for the whole bunch of methods. I now did the latter, I grouped the methods that can be implemented using SLEEF together and used a single `if defined ... else` block.

Sounds good. I will take a look at it Monday if thats ok.

Regarding implementation: THis might be possible but not quite sure.
```
#define USE_SLEEF_ON_MAC(sleef_code, non_sleef_code) \
  #ifdef.....
     sleef_code
  #else
    non_sleef_code
  #endif 


... exp(...) {
  USE_SLEEF_ON_MAC(
    {
      Sleef....
    },
    {
      map(std::exp)...
    })
}
```

**Comment by malfet** [comment]
Smaller PR looks good to me, though I'm not sure I understand the argument why not use sleef unconditionally for ARM platform when it is available?
Agree with @kimishpatel suggestion about unified macro rather than one huge ifdef

**Comment by danieldk** [comment]
> Smaller PR looks good to me, though I'm not sure I understand the argument why not use sleef unconditionally for ARM platform when it is available?

Makes sense. I think the comment at the top of the source file gave me the impression that for some reason it shouldn't be used on every AArch64 platform.

> Agree with @kimishpatel suggestion about unified macro rather than one huge ifdef

Ack.

**Comment by danieldk** [comment]
Rewritten with a `USE_SLEEF` macro as @kimishpatel suggested. I think it looks more clean now, except it gets a bit whacky for multi-line functions, but I guess that's C macros 🙈.

File: BUILD.bazel [hunk 1]
```diff
@@ -320,6 +320,7 @@ header_template_rule(
         "@CAFFE2_STATIC_LINK_CUDA_INT@": "0",
         "@AT_BUILD_WITH_BLAS@": "1",
         "@AT_BUILD_WITH_LAPACK@": "1",
+        "@AT_BUILD_WITH_SLEEF@": "1",
```
Comment (kimishpatel) [inline BUILD.bazel:323]: These macros generally apply for pytorch codebase. In this case we are trying to use it for specific implementation of an interface. From the code it seems that it is not really used for building sleef but more about using it to 1) include <sleef.h> and 2) use the sleef-code.

So maybe rename this as `AT_BUILD_ARM_VEC256_WITH_SLEEF` (sorry a mouthful) and set it to true by default. That way sleef is used for all oss arm64 builds. For internal fb stuff we will revisit it. @malfet to confirm if thats what he had in mind.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Sorry for delay in reviewing this. Left some comments.

File: aten/src/ATen/cpu/vec/vec256/vec256_float_neon.h [hunk 2]
```diff
@@ -324,68 +335,105 @@ template <> class Vectorized<float> {
...
-    __at_align__ float tmp_exp[size()];
-    store(tmp);
-    exp.store(tmp_exp);
-    for (const auto i : c10::irange(size())) {
-      tmp[i] = std::atan2(tmp[i], tmp_exp[i]);
-    }
-    return loadu(tmp);
+    USE_SLEEF(
+      return Vectorized<float>(Sleef_atan2f4_u10(values.val[0], exp.values.val[0]),
+                               Sleef_atan2f4_u10(values.val[1], exp.values.val[1])),
```
[the remark this answers is not in the document]
Comment (danieldk) [reply aten/src/ATen/cpu/vec/vec256/vec256_float_neon.h]: Done ✅

File: BUILD.bazel (hunk 1 above)
Comment (danieldk) [reply BUILD.bazel]: Renamed and I have set it to be always enabled.

**Comment by kimishpatel** [comment]
Imported. Will see if any additional changes needed if the changes break anything.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Need to make some changes to make it work for internal builds

File: aten/src/ATen/CMakeLists.txt [hunk 3]
```diff
@@ -32,6 +32,7 @@ set_bool(AT_BLAS_F2C BLAS_F2C)
 set_bool(AT_BLAS_USE_CBLAS_DOT BLAS_USE_CBLAS_DOT)
 set_bool(AT_MAGMA_ENABLED USE_MAGMA)
 set_bool(CAFFE2_STATIC_LINK_CUDA_INT CAFFE2_STATIC_LINK_CUDA)
+set_bool(AT_BUILD_ARM_VEC256_WITH_SLEEF 1)
```
[the remark this answers is not in the document]
Comment (danieldk) [reply aten/src/ATen/CMakeLists.txt]: No problem! I have added a commit that adds the CMake option that you suggest.

**Review by kimishpatel — APPROVED** [review]
I think this worked. I will work on landing it.

**Comment by danieldk** [comment]
Thanks for merging this! Do I have to do anything to get this enabled in default macOS M1 builds (e.g. a pull request against `pytorch/builder`)?

**Comment by wojciak** [comment]
Hi all :) 

This is really interesting! Was this closed without a merge though?

**Comment by danieldk** [comment]
> This is really interesting! Was this closed without a merge though?

It was merged. If you look at https://github.com/pytorch/pytorch/commit/d50211860ab3816ea563ad105930c62d8142cf60, you can see that it is on the `master` branch.
